go/pkg/services/logcollector/server.go

```diff
@@ -0,0 +1,605 @@
+}
+
+func (s *LogCollectorServer) hasLogs(ctx context.Context, runUID string, streamReader *bufio.Reader) bool {
+
+	// peek into the stream, and wait until there is something to read from it
+	// or until context is canceled
+	for {
+		select {
+		case <-time.After(s.readLogWaitTime):
+			peekBuf, err := streamReader.Peek(1)
+
+			// if there is something to read, return true
+			// if error is EOF, the pod has logs but not new ones
+			if err == io.EOF || len(peekBuf) > 0 {
+				return true
+			}
+			if err != nil {
+				s.Logger.WarnWithCtx(ctx, "Failed to peek into stream", "runUID", runUID, "err", err.Error())
+			}
+		case <-ctx.Done():
+			s.Logger.DebugWithCtx(ctx, "Context was canceled, stopping waiting for pod log stream", "runUID", runUID)
+			return false
+		}
+		time.Sleep(s.readLogWaitTime)
+	}
+}
+
+func (s *LogCollectorServer) cancelOnContext(ctx context.Context, file *os.File, copyBufferDone chan struct{}) {
+	select {
+	case <-ctx.Done():
+
+		// context is dead, so we close the file so `CopyBuffer` won't block and fail
+		file.Close() // nolint: errcheck
+	case <-copyBufferDone:
+
+		// `CopyBuffer` doesn't block anymore, we can stop the goroutine
+		return
+	}
+}
+
+func (s *LogCollectorServer) getLogFilePath(runUID string) (string, error) {
+
+	logFilePath := ""
+	var latestModTime time.Time
+
+	// list all files in base directory
+	err := filepath.Walk(s.baseDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		// if file name starts with run id, it's a log file
+		if strings.HasPrefix(info.Name(), runUID) {
+
+			// if it's the first file, set it as the log file path
+			// otherwise, check if it's the latest modified file
+			if logFilePath == "" || info.ModTime().After(latestModTime) {
+				logFilePath = path
+				latestModTime = info.ModTime()
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		return "", errors.Wrap(err, "Failed to list files in base directory")
+	}
+
+	// if no log file was found, return error
+	if logFilePath == "" {
+		return "", errors.Errorf("Failed to find log file for run id %s", runUID)
+	}
+
+	return logFilePath, nil
+}
+
+func (s *LogCollectorServer) readLogsFromFile(ctx context.Context,
+	runUID,
+	filePath string,
+	offset uint64,
+	size int64) ([]byte, error) {
+
+	// open log file for reading
+	file, err := os.OpenFile(filePath, os.O_RDONLY, 0644)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to open log file for run id %s", runUID)
+	}
+	defer file.Close() // nolint: errcheck
+
+	// get file size
+	fileInfo, err := file.Stat()
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to get file info for run id %s", runUID)
+	}
+
+	offset, size = s.validateOffsetAndSize(offset, size, fileInfo.Size())
+	if size == 0 {
+		s.Logger.DebugWithCtx(ctx, "No logs to return", "run_id", runUID)
+		return nil, nil
+	}
+
+	// read size bytes from offset
+	buffer := make([]byte, size)
```

Comment:
why not buffer pool here? remember, 1000 runs, 10k requests to get logs, 10k buffers, each size lets say of 32kb can get to 320MB.
on top of that, think of how many objects are being collected and GCed. perhaps you can recycle some objects and reuse?

Reply:
If I get a request for `size` bytes, I will have to keep them somewhere.
Say I'm using the constant-sized buffers from the pool to read from the file, I will need to return them as a byte slice with size `size`, which is the concatenation of all buffers used for reading.
How can I reuse buffers of varying sizes then?